In [1]:
from sqlalchemy import create_engine
import os

def create_connection():
    host = 'localhost'
    port = '3333'
    return create_engine(f"postgresql://infocup@{host}:{port}/infocup")

# Aufbau Tankstellen Daten

Für das Clustering der Tankstellen ist wichtig welche Tankstelle in welchem  Zeitraum existiert.

In [2]:
import pandas as pd
con = create_connection()

df = pd.read_sql_query("""
select id, min_ts, max_ts
from stations
where min_ts is not null""", con=con)
df



id              min_ts              max_ts
0       4554 2014-06-08 12:38:01 2017-09-22 02:07:06
1       4690 2014-06-09 00:02:01 2017-09-21 22:26:05
2       6897 2014-06-08 14:02:01 2017-09-21 22:01:07
3       8290 2014-06-08 12:14:01 2017-09-21 23:10:05
4       8404 2014-06-09 00:02:01 2017-09-22 00:04:05
5       8908 2014-06-08 11:18:01 2016-11-03 23:04:31
6       9407 2014-06-09 00:02:01 2017-09-21 21:57:06
7       9571 2017-04-27 12:04:31 2017-09-21 23:03:06
8      10011 2014-06-08 13:18:01 2017-09-21 23:08:05
9      10495 2014-06-09 00:02:01 2017-09-21 22:01:07
10     10553 2014-06-09 00:02:01 2017-09-19 23:16:20
11     11361 2014-06-09 00:02:01 2017-09-21 21:57:06
12     13491 2014-06-10 11:14:01 2017-09-21 14:37:22
13     14089 2014-06-09 00:02:01 2017-09-21 23:05:06
14     14682 2014-06-09 00:02:01 2017-09-21 23:08:05
15      1018 2014-06-10 02:50:01 2017-09-21 23:22:05
16      3745 2014-06-08 11:14:01 2017-09-21 22:01:07
17      3746 2014-06-08 10:46:01 2017-09-21 23:10:05
18      3747 2014-06-08 10:26:01 2017-09-21 23:03:06
19      3748 2014-06-09 00:02:01 2017-09-21 23:18:05
20      3749 2014-06-08 10:22:01 2017-09-21 22:01:07
21      3750 2014-06-08 10:18:01 2017-09-21 22:57:06
22      3751 2014-06-09 04:22:01 2017-09-21 21:05:06
23      3752 2014-06-08 17:06:01 2017-09-21 23:03:06
24      3753 2014-06-09 02:10:01 2017-09-21 23:14:05
25      3754 2014-06-09 00:02:01 2017-09-21 19:48:06
26      3755 2014-06-08 09:58:01 2016-01-04 12:02:01
27      3756 2016-01-04 16:06:01 2017-09-22 01:19:05
28      3757 2014-06-08 10:50:01 2017-09-21 23:03:06
29      3758 2014-06-09 00:02:01 2017-09-21 22:01:07
...      ...                 ...                 ...
14927  15197 2014-06-09 00:02:01 2017-09-21 22:04:24
14928  15198 2014-06-09 00:02:01 2017-09-21 21:54:12
14929  15199 2014-06-10 07:14:01 2017-09-21 22:04:24
14930  15200 2014-06-17 05:42:01 2017-09-21 08:03:06
14931  15201 2014-06-10 09:02:01 2017-09-21 22:24:05
14932  15202 2015-10-13 18:26:01 2017-09-19 23:16:20
14933  15203 2014-06-09 00:02:01 2017-09-21 22:04:24
14934  15204 2014-06-09 18:10:01 2017-09-20 08:24:06
14935  15205 2014-06-09 00:02:01 2017-09-21 22:01:07
14936  15206 2014-06-18 11:30:01 2017-09-21 10:02:06
14937  15207 2014-06-10 01:58:01 2017-09-21 21:05:06
14938  15208 2014-06-25 11:02:01 2016-06-13 14:02:01
14939  15209 2014-06-09 00:02:01 2017-09-21 23:10:05
14940  15210 2014-06-09 00:02:01 2017-09-20 12:09:06
14941  15211 2017-04-25 14:35:31 2017-09-21 23:16:05
14942  15212 2014-06-08 14:02:01 2017-09-21 22:01:07
14943  15213 2014-06-09 00:02:01 2017-09-21 23:25:06
14944  15214 2014-06-09 00:02:01 2017-09-21 22:01:07
14945  15215 2014-06-08 10:38:01 2017-09-21 23:03:06
14946  15216 2014-06-09 00:02:01 2017-09-21 22:04:24
14947  15217 2014-06-09 00:02:01 2017-09-21 23:10:05
14948  15218 2014-06-08 14:02:01 2017-09-21 22:01:07
14949  15219 2014-06-09 00:02:01 2017-09-21 23:03:06
14950  15220 2014-06-09 00:02:01 2017-09-21 22:04:24
14951  15221 2014-06-09 00:02:01 2017-09-21 23:03:06
14952  15222 2014-06-09 01:42:01 2017-09-20 09:38:05
14953  15223 2014-06-08 10:14:01 2017-09-21 23:10:05
14954  15224 2014-06-09 00:02:01 2017-09-21 23:03:06
14955  15225 2016-12-01 12:56:31 2017-09-21 22:01:07
14956  15226 2014-06-12 12:54:01 2016-10-31 12:24:31

[14957 rows x 3 columns]

In [3]:
min_tss = df['min_ts'].copy()
min_tss = pd.Series(min_tss.values.astype('<M8[D]'))
min_tss.groupby(min_tss).count()

2014-06-08    5781
2014-06-09    6468
2014-06-10     904
2014-06-11      75
2014-06-12      67
2014-06-13      38
2014-06-14      18
2014-06-15      11
2014-06-16      23
2014-06-17      29
2014-06-18      31
2014-06-19       7
2014-06-20      29
2014-06-21       2
2014-06-22       6
2014-06-23      14
2014-06-24       5
2014-06-25      13
2014-06-26       8
2014-06-27       4
2014-06-28       1
2014-06-29       1
2014-06-30       6
2014-07-01       4
2014-07-02       6
2014-07-03      13
2014-07-04       4
2014-07-05       1
2014-07-06       3
2014-07-07       3
              ... 
2017-07-17       1
2017-07-18       1
2017-07-20       1
2017-07-22       4
2017-07-27       1
2017-08-01       1
2017-08-02       1
2017-08-07       1
2017-08-08       5
2017-08-09       2
2017-08-11       1
2017-08-14       2
2017-08-15       4
2017-08-16       1
2017-08-18       2
2017-08-21       1
2017-08-22       2
2017-08-23       2
2017-08-25       2
2017-08-29       3
2017-08-30       2
2017-08-31  

In [4]:
max_tss = df['max_ts'].copy()
max_tss = pd.Series(max_tss.values.astype('<M8[D]'))
max_tss.groupby(max_tss).count()

2014-08-02        1
2014-08-15        3
2014-10-13        1
2014-11-28        1
2014-12-31        1
2015-04-27        1
2015-06-01        1
2015-06-02        1
2015-06-29        1
2015-08-04        1
2015-08-14        1
2015-09-01        1
2015-09-30        1
2015-10-01        1
2015-10-05        1
2015-11-24        1
2015-12-01        1
2015-12-11        1
2015-12-20        1
2016-01-04        1
2016-01-06        1
2016-02-03        1
2016-02-29        1
2016-03-17        1
2016-03-30        1
2016-03-31        2
2016-04-10        1
2016-04-19        1
2016-04-22        2
2016-05-01        7
              ...  
2017-07-21        1
2017-07-24        1
2017-07-25        1
2017-07-26        1
2017-07-28       17
2017-07-29        1
2017-07-31        8
2017-08-01        2
2017-08-02        2
2017-08-07        1
2017-08-14        1
2017-08-15        1
2017-08-18        1
2017-08-21        1
2017-08-24       10
2017-08-28        1
2017-08-29        1
2017-08-30        2
2017-08-31        4


Zwischen 2015-01-01 und 2017-09-18 sollten genug Tankstellen durchgehend Daten haben, um weitere Analysen durchzuführen.

In [5]:
active_stations = pd.read_sql_query("""
select id, min_ts, max_ts
from stations
where min_ts <= '2015-01-01'
and max_ts >= '2017-09-18'""", con)
active_stations

id              min_ts              max_ts
0       4554 2014-06-08 12:38:01 2017-09-22 02:07:06
1       4690 2014-06-09 00:02:01 2017-09-21 22:26:05
2       6897 2014-06-08 14:02:01 2017-09-21 22:01:07
3       8290 2014-06-08 12:14:01 2017-09-21 23:10:05
4       8404 2014-06-09 00:02:01 2017-09-22 00:04:05
5       9407 2014-06-09 00:02:01 2017-09-21 21:57:06
6      10011 2014-06-08 13:18:01 2017-09-21 23:08:05
7      10495 2014-06-09 00:02:01 2017-09-21 22:01:07
8      10553 2014-06-09 00:02:01 2017-09-19 23:16:20
9      11361 2014-06-09 00:02:01 2017-09-21 21:57:06
10     13491 2014-06-10 11:14:01 2017-09-21 14:37:22
11     14089 2014-06-09 00:02:01 2017-09-21 23:05:06
12     14682 2014-06-09 00:02:01 2017-09-21 23:08:05
13      1018 2014-06-10 02:50:01 2017-09-21 23:22:05
14      3745 2014-06-08 11:14:01 2017-09-21 22:01:07
15      3746 2014-06-08 10:46:01 2017-09-21 23:10:05
16      3747 2014-06-08 10:26:01 2017-09-21 23:03:06
17      3748 2014-06-09 00:02:01 2017-09-21 23:18:05
18      3749 2014-06-08 10:22:01 2017-09-21 22:01:07
19      3750 2014-06-08 10:18:01 2017-09-21 22:57:06
20      3751 2014-06-09 04:22:01 2017-09-21 21:05:06
21      3752 2014-06-08 17:06:01 2017-09-21 23:03:06
22      3753 2014-06-09 02:10:01 2017-09-21 23:14:05
23      3754 2014-06-09 00:02:01 2017-09-21 19:48:06
24      3757 2014-06-08 10:50:01 2017-09-21 23:03:06
25      3758 2014-06-09 00:02:01 2017-09-21 22:01:07
26      3759 2014-06-10 02:34:01 2017-09-22 00:08:06
27      3760 2014-06-09 00:02:01 2017-09-21 23:15:06
28      3761 2014-06-08 10:42:01 2017-09-21 22:17:06
29      3762 2014-06-08 09:54:01 2017-09-21 22:01:07
...      ...                 ...                 ...
13585  15191 2014-06-08 10:38:01 2017-09-21 23:03:06
13586  15193 2014-06-09 00:02:01 2017-09-21 22:04:24
13587  15194 2014-06-09 00:02:01 2017-09-21 23:10:05
13588  15195 2014-06-09 00:02:01 2017-09-21 22:04:24
13589  15196 2014-06-08 15:18:01 2017-09-21 22:23:05
13590  15197 2014-06-09 00:02:01 2017-09-21 22:04:24
13591  15198 2014-06-09 00:02:01 2017-09-21 21:54:12
13592  15199 2014-06-10 07:14:01 2017-09-21 22:04:24
13593  15200 2014-06-17 05:42:01 2017-09-21 08:03:06
13594  15201 2014-06-10 09:02:01 2017-09-21 22:24:05
13595  15203 2014-06-09 00:02:01 2017-09-21 22:04:24
13596  15204 2014-06-09 18:10:01 2017-09-20 08:24:06
13597  15205 2014-06-09 00:02:01 2017-09-21 22:01:07
13598  15206 2014-06-18 11:30:01 2017-09-21 10:02:06
13599  15207 2014-06-10 01:58:01 2017-09-21 21:05:06
13600  15209 2014-06-09 00:02:01 2017-09-21 23:10:05
13601  15210 2014-06-09 00:02:01 2017-09-20 12:09:06
13602  15212 2014-06-08 14:02:01 2017-09-21 22:01:07
13603  15213 2014-06-09 00:02:01 2017-09-21 23:25:06
13604  15214 2014-06-09 00:02:01 2017-09-21 22:01:07
13605  15215 2014-06-08 10:38:01 2017-09-21 23:03:06
13606  15216 2014-06-09 00:02:01 2017-09-21 22:04:24
13607  15217 2014-06-09 00:02:01 2017-09-21 23:10:05
13608  15218 2014-06-08 14:02:01 2017-09-21 22:01:07
13609  15219 2014-06-09 00:02:01 2017-09-21 23:03:06
13610  15220 2014-06-09 00:02:01 2017-09-21 22:04:24
13611  15221 2014-06-09 00:02:01 2017-09-21 23:03:06
13612  15222 2014-06-09 01:42:01 2017-09-20 09:38:05
13613  15223 2014-06-08 10:14:01 2017-09-21 23:10:05
13614  15224 2014-06-09 00:02:01 2017-09-21 23:03:06

[13615 rows x 3 columns]

13615 Tankstellen sind zu viel, um verschieden Clustering Verfahren miteinander zu vergleichen und noch dieses Jahr ein Ergebnis zu erwarten. Deshalb werden die Ids der in Frage kommenden Tankstellen gesampled.

In [6]:
import numpy as np
ids = np.random.choice(active_stations['id'], 100)
ids

array([ 9878,  8368,  5353,  6923,  1196, 14317,  2848, 11156, 11560,
        4351,  7006,  3975,  8372,  4405,   947,  4491,  3650, 14578,
        9782,   173,  4807, 11295,  9026, 14240, 13293, 11629,  5884,
       11055, 13599,  3382,  3445, 13360, 14758, 15121, 12392, 13260,
        3864,   112,  4893,   233,  4820, 14859,  2387,  6774, 11348,
        1384,  6249,  5200, 14610,  6599, 14403, 12805,   588,  6316,
        4665,  5397, 10867,  3167, 10292,  3989, 12127,  4883, 12605,
        5810,  1277, 14459,  6117,   829,  2866, 14374,  3396,  8408,
       10742,  5145,  7959,  1900, 11048, 12480,  1890, 11342,  3973,
        8539,  2408, 13447, 12285, 10353,  7114, 11395,  2927, 10528,
       11143,  3507,  8398,  7609,   957,  4470, 12044,  4141,  2145,  8020], dtype=int64)

In [57]:
param = ','.join([str(x) for x in ids])
query = """select * from prices_sampled where station_id in (%s) 
and time_stamp between '2015-01-01' and '2017-09-18'""" % param
prices = pd.read_sql_query(query, con)
prices


time_stamp  price  station_id
0      2015-01-01   1389         957
1      2015-01-01   1319        1890
2      2015-01-01   1414        1900
3      2015-01-01   1299        2145
4      2015-01-01   1319        2408
5      2015-01-01   1339        2927
6      2015-01-01   1309        3396
7      2015-01-01   1329        3507
8      2015-01-01   1379        3973
9      2015-01-01   1329        4141
10     2015-01-01   1399        4470
11     2015-01-01   1359        5145
12     2015-01-01   1359        7114
13     2015-01-01   1279        7609
14     2015-01-01   1359        7959
15     2015-01-01   1519        8020
16     2015-01-01   1389        8398
17     2015-01-01   1379        8408
18     2015-01-01   1349        8539
19     2015-01-01   1349       10353
20     2015-01-01   1339       10528
21     2015-01-01   1389       10742
22     2015-01-01   1379       11048
23     2015-01-01   1349       11143
24     2015-01-01   1339       11342
25     2015-01-01   1289       11395
26     2015-01-01   1269       12044
27     2015-01-01   1379       12285
28     2015-01-01   1409       12480
29     2015-01-01   1339       13447
...           ...    ...         ...
713490 2017-09-18   1419         957
713491 2017-09-18   1389        1890
713492 2017-09-18   1554        1900
713493 2017-09-18   1329        2145
713494 2017-09-18   1369        2408
713495 2017-09-18   1369        2927
713496 2017-09-18   1329        3396
713497 2017-09-18   1419        3507
713498 2017-09-18   1439        3973
713499 2017-09-18   1389        4141
713500 2017-09-18   1489        4470
713501 2017-09-18   1459        5145
713502 2017-09-18   1429        7114
713503 2017-09-18   1409        7609
713504 2017-09-18   1489        7959
713505 2017-09-18   1419        8020
713506 2017-09-18   1429        8398
713507 2017-09-18   1479        8408
713508 2017-09-18   1399        8539
713509 2017-09-18   1419       10353
713510 2017-09-18   1379       10528
713511 2017-09-18   1429       10742
713512 2017-09-18   1439       11048
713513 2017-09-18   1419       11143
713514 2017-09-18   1419       11342
713515 2017-09-18   1309       11395
713516 2017-09-18   1319       12044
713517 2017-09-18   1489       12285
713518 2017-09-18   1489       12480
713519 2017-09-18   1389       13447

[713520 rows x 3 columns]

In [67]:
prices_pivoted = prices.pivot_table(index='time_stamp', columns='station_id', values='price')
prices_pivoted

station_id           957    1890   1900   2145   2408   2927   3396   3507   \
time_stamp                                                                    
2015-01-01 00:00:00   1389   1319   1414   1299   1319   1339   1309   1329   
2015-01-01 01:00:00   1389   1319   1414   1299   1319   1339   1309   1329   
2015-01-01 02:00:00   1389   1319   1414   1299   1319   1339   1309   1329   
2015-01-01 03:00:00   1389   1319   1414   1299   1319   1339   1309   1329   
2015-01-01 04:00:00   1389   1319   1414   1299   1319   1339   1309   1329   
2015-01-01 05:00:00   1389   1319   1414   1299   1319   1339   1309   1329   
2015-01-01 06:00:00   1329   1319   1414   1299   1319   1339   1309   1329   
2015-01-01 07:00:00   1329   1309   1414   1299   1319   1339   1309   1329   
2015-01-01 08:00:00   1329   1309   1414   1299   1319   1339   1309   1329   
2015-01-01 09:00:00   1329   1299   1414   1299   1319   1339   1309   1329   
2015-01-01 10:00:00   1329   1289   1414   1299   1319   1339   1309   1329   
2015-01-01 11:00:00   1329   1289   1374   1299   1319   1339   1309   1329   
2015-01-01 12:00:00   1329   1289   1374   1299   1319   1339   1309   1279   
2015-01-01 13:00:00   1329   1289   1334   1299   1319   1339   1309   1279   
2015-01-01 14:00:00   1329   1289   1334   1299   1319   1339   1309   1279   
2015-01-01 15:00:00   1319   1289   1334   1299   1319   1339   1309   1279   
2015-01-01 16:00:00   1319   1279   1334   1299   1319   1339   1309   1269   
2015-01-01 17:00:00   1289   1279   1334   1299   1319   1339   1309   1269   
2015-01-01 18:00:00   1289   1279   1324   1299   1319   1339   1309   1269   
2015-01-01 19:00:00   1269   1279   1314   1299   1319   1339   1309   1269   
2015-01-01 20:00:00   1269   1279   1314   1299   1319   1339   1309   1269   
2015-01-01 21:00:00   1269   1279   1344   1299   1319   1339   1309   1269   
2015-01-01 22:00:00   1369   1279   1344   1299   1319   1339   1309   1269   
2015-01-01 23:00:00   1369   1279   1344   1299   1319   1339   1309   1269   
2015-01-02 00:00:00   1389   1339   1404   1299   1319   1339   1309   1329   
2015-01-02 01:00:00   1389   1339   1404   1299   1319   1339   1309   1329   
2015-01-02 02:00:00   1389   1339   1404   1299   1319   1339   1309   1329   
2015-01-02 03:00:00   1389   1339   1404   1299   1319   1339   1309   1329   
2015-01-02 04:00:00   1389   1339   1404   1299   1319   1339   1309   1329   
2015-01-02 05:00:00   1389   1339   1404   1299   1319   1339   1309   1329   
...                    ...    ...    ...    ...    ...    ...    ...    ...   
2017-09-16 19:00:00   1299   1299   1394   1319   1379   1369   1339   1299   
2017-09-16 20:00:00   1299   1289   1394   1319   1379   1369   1309   1299   
2017-09-16 21:00:00   1299   1289   1394   1319   1379   1369   1309   1299   
2017-09-16 22:00:00   1299   1289   1394   1299   1379   1369   1309   1299   
2017-09-16 23:00:00   1299   1289   1394   1299   1379   1369   1309   1299   
2017-09-17 00:00:00   1419   1389   1554   1329   1379   1369   1309   1389   
2017-09-17 01:00:00   1419   1389   1554   1329   1379   1369   1309   1389   
2017-09-17 02:00:00   1419   1389   1554   1329   1379   1369   1309   1389   
2017-09-17 03:00:00   1419   1389   1554   1329   1379   1369   1309   1389   
2017-09-17 04:00:00   1419   1389   1554   1329   1379   1369   1309   1389   
2017-09-17 05:00:00   1419   1389   1554   1329   1379   1369   1309   1389   
2017-09-17 06:00:00   1419   1389   1554   1329   1379   1469   1309   1389   
2017-09-17 07:00:00   1399   1389   1474   1389   1379   1469   1309   1389   
2017-09-17 08:00:00   1399   1389   1474   1389   1439   1469   1309   1389   
2017-09-17 09:00:00   1399   1389   1474   1389   1399   1469   1399   1389   
2017-09-17 10:00:00   1399   1319   1474   1329   1379   1399   1389   1369   
2017-09-17 11:00:00   1359   1309   1474   1329   1379   1319   1369   1349   
2017-09-17 12:00:00   1339   1299   1434   1329   